be/src/exec/pipeline/scan/morsel.cpp

```diff
@@ -229,5 +238,315 @@ Status PhysicalSplitMorselQueue::_init_segment() {
     return Status::OK();
 }
 
+std::vector<TInternalScanRange*> LogicalSplitMorselQueue::olap_scan_ranges() const {
+    return _convert_morsels_to_olap_scan_ranges(_morsels);
+}
+
+void LogicalSplitMorselQueue::set_key_ranges(const std::vector<std::unique_ptr<OlapScanRange>>& key_ranges) {
+    for (const auto& key_range : key_ranges) {
+        if (key_range->begin_scan_range.size() == 1 && key_range->begin_scan_range.get_value(0) == NEGATIVE_INFINITY) {
+            continue;
+        }
+
+        _range_start_op = key_range->begin_include ? vectorized::TabletReaderParams::RangeStartOperation::GE
+                                                   : vectorized::TabletReaderParams::RangeStartOperation::GT;
+        _range_end_op = key_range->end_include ? vectorized::TabletReaderParams::RangeEndOperation::LE
+                                               : vectorized::TabletReaderParams::RangeEndOperation::LT;
+
+        _range_start_key.emplace_back(key_range->begin_scan_range);
+        _range_end_key.emplace_back(key_range->end_scan_range);
+    }
+}
+
+StatusOr<MorselPtr> LogicalSplitMorselQueue::try_get() {
+    DCHECK(!_tablets.empty());
+    DCHECK(!_tablet_rowsets.empty());
+    DCHECK_EQ(_tablets.size(), _tablet_rowsets.size());
+
+    std::lock_guard<std::mutex> lock(_mutex);
+
+    if (_tablet_idx >= _tablets.size()) {
+        return nullptr;
+    }
+
+    // When it hasn't initialized any tablet,
+    // or the current tablet doesn't contain any segment,
+    // or all the key ranges of the current tablet has been finished,
+    // we should pick up the next tablet and init it.
+    while (!_has_init_any_tablet || _segment_group == nullptr || _cur_tablet_finished()) {
+        if (!_next_tablet()) {
+            return nullptr;
+        }
+        RETURN_IF_ERROR(_init_tablet());
+    }
+
+    size_t num_taken_blocks = 0;
+    std::vector<vectorized::ShortKeyRangeOptionPtr> short_key_ranges;
+    vectorized::ShortKeyOptionPtr _cur_range_lower = nullptr;
+    vectorized::ShortKeyOptionPtr _cur_range_upper = nullptr;
+    bool need_more_blocks = true;
+    while (!_cur_tablet_finished() &&      // One morsel only read data from one tablet.
```

Comment:
Had better to give some examples to illustrate the detail

Reply:
Great.